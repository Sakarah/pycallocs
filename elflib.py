import importlib
import importlib.abc
import importlib.machinery
import sys
# Re-export everything from the C extension module
from allocs import *

# These can be modified by the user to change the finder search paths
lib_extension = ".so"

# The blank search path has a special meaning as dlopen is using its own search 
# paths when the file name does not contain any /
__path__ = ["./", ""]

class LibraryFinder(importlib.abc.MetaPathFinder):
    """
        Meta path finder searching foreign libraries usable
        with allocs.ForeignLibraryLoader.
    """

    def find_spec(fullname, path, target):
        if path is None:
            # We are defining __path__ so it should never be None
            return None

        # We only try to load modules below ourselves
        modpath = fullname.split(".")
        if len(modpath) != 2 or modpath[0] != __name__:
            return None
        name = modpath[1]

        for base_path in path:
            try:
                filename = base_path + name + lib_extension
                loader = LibraryLoader(filename)
                return importlib.machinery.ModuleSpec(fullname, loader, origin=filename)
            except ImportError:
                continue

        return None

# We must register ourselves before the default meta path finder to prevent exception
# ImportError: dynamic module does not define module export function (PyInit_%)
sys.meta_path.insert(0, LibraryFinder)
