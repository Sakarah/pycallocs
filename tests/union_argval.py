import elflib
elflib.__path__.append("libs/")
from elflib import composite as m

a = m.quantum_cat()
a.dead = 42
m.print_cat(a)

a.alive = 9.99
m.print_cat(a)
