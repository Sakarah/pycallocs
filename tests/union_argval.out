Dead (42) or alive (0.00)
Dead (1202590843) or alive (9.99)
